packages/protocol/analysisManager.ts

```diff
@@ -166,7 +167,10 @@ class AnalysisManager {
       this.handlers.set(analysisId, handler);
       await Promise.all([
         !handler.onAnalysisResult || client.Analysis.runAnalysis({ analysisId }, sessionId),
-        !handler.onAnalysisPoints || client.Analysis.findAnalysisPoints({ analysisId }, sessionId),
+        !(handler.onAnalysisPoints || handler.onPointsFinished) ||
+          client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
+            handler.onPointsFinished
+          ),
```

Comment:
This looks wrong, I think.

If you pass both `onAnalysisPoints` but not `onPointsFinished` – this will add a handler for `onPointsFinished`?

Reply:
The `onAnalysisPoints()` handler will be called by the `AnalysisManager.onAnalysisPoints()` method which receives the `Analysis.analysisPoints` events. But the backend will only send these events to us if we call the `Analysis.findAnalysisPoints` command, that's why the `onAnalysisPoints()` handler appears in this condition.

Reply:
I don't think I understand. Those things seem kind of far apart from each other to be an uncommented dependency in this function. (Also maybe it's just me but I really have trouble parsing the way this code is written.)

Could we write it without as much `!` and `||` to maybe make it more clearer?
Is this the same code?

```js
const promises: Promise<any>[] = [];
if (handler.onAnalysisResult) {
  promises.push(client.Analysis.runAnalysis({ analysisId }, sessionId));
}
if (handler.onAnalysisPoints && handler.onPointsFinished) {
  promises.push(
    client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
      handler.onPointsFinished
    )
  );
}
await Promise.all(promises);
```

Reply:
Yes, this is the same code except we still need to say `if (handler.onAnalysisPoints || handler.onPointsFinished)` instead of `if (handler.onPointsFinished)`.

Reply:
Can you replace the above with something like this then, for readability?
```js
const promises: Promise<any>[] = [];
if (handler.onAnalysisResult) {
  promises.push(client.Analysis.runAnalysis({ analysisId }, sessionId));
}
if (handler.onAnalysisPoints || handler.onPointsFinished) {
  // We check onAnalysisPoints too even though it isn't directly referenced here
  // because unless we call Analysis.findAnalysisPoints,
  // that callback won't be called.
  promises.push(
    client.Analysis.findAnalysisPoints({ analysisId }, sessionId).then(
      handler.onPointsFinished
    )
  );
}
await Promise.all(promises);
```